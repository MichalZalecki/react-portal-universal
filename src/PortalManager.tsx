import * as React from 'react';

export type ChildrenSelectorTuple = [any, string];

const PortalContext = React.createContext<ChildrenSelectorTuple[]>([]);

export const PortalConsumer = PortalContext.Consumer;

export const PortalManager: React.SFC<{ portals: ChildrenSelectorTuple[] }> = ({ children, portals }) => (
  <PortalContext.Provider value={portals}>
    {children}
  </PortalContext.Provider>
);
