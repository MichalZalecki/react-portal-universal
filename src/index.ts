import * as ReactDOM from "react-dom";

export type ChildrenSelectorTuple = [any, string];

const portals: ChildrenSelectorTuple[] = [];

function canUseDOM() {
  return !!(typeof window !== 'undefined' && window.document && window.document.createElement);
}

export function createUniversalPortal(children: any, selector: string) {
  if (!canUseDOM()) {
    portals.push([children, selector]); // yes, mutation (҂◡_◡)
    return null;                        // do not render anything on the server
  }
  // TODO: Do not cast to any when typings are updated for createPortal
  return (ReactDOM as any).createPortal(children, document.querySelector(selector));
}

export function flushUniversalPortals(): ChildrenSelectorTuple[] {
  const copy = portals.slice();
  portals.length = 0;                   // it's important to flush one and only one time per render
  return copy;
}

export function removeUniversalPortals() {
  if (canUseDOM()) {
    Array.prototype.slice.call(document.querySelectorAll("[data-react-universal-portal]")).forEach(function (node) {
      node.remove();
    });
  }
}
