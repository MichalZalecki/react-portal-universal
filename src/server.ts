import * as ReactDOMServer from "react-dom/server";
import { load }  from "cheerio";
import { flushUniversalPortals } from "./index";

export function appendUniversalPortals(html: string) {
  const portals = flushUniversalPortals();
  if (!portals.length) {
    return html;
  }
  const $ = load(html);
  portals.forEach(([children, selector]) => {
    const markup = ReactDOMServer.renderToStaticMarkup(children);
    $(markup).attr("data-react-universal-portal", "").appendTo((selector as any))
  });
  return $.html({ decodeEntities: false });
}
