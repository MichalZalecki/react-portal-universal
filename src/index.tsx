import * as React from 'react';
import * as ReactDOM from "react-dom";
import { PortalConsumer, ChildrenSelectorTuple } from "./PortalManager";

function canUseDOM() {
  return !!(typeof window !== 'undefined' && window.document && window.document.createElement);
}

function createUniversalPortal(children: React.ReactNode, selector: string, portals: ChildrenSelectorTuple[]) {
  if (!canUseDOM()) {
    portals.push([children, selector]); // yes, mutation (҂◡_◡)
    return null;                        // do not render anything on the server
  }
  // TODO: Do not cast to any when typings are updated for createPortal
  return (ReactDOM as any).createPortal(children, document.querySelector(selector));
}

export function prepareClientPortals() {
  if (canUseDOM()) {
    Array.prototype.slice.call(document.querySelectorAll("[data-react-universal-portal]")).forEach(function (node: Element) {
      node.remove();
    });
  }
}

export const UniversalPortal: React.SFC<{ selector: string }> = ({ children, selector }) => (
  <PortalConsumer>
    {(portals: ChildrenSelectorTuple[]) => createUniversalPortal(children, selector, portals)}
  </PortalConsumer>
);
