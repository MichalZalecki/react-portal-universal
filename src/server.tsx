import * as React from "react";
import * as ReactDOMServer from "react-dom/server";
import { load }  from "cheerio";
import { PortalManager, ChildrenSelectorTuple } from "./PortalManager";

export class ServerPortal {
  portals: ChildrenSelectorTuple[];

  constructor() {
    this.portals = [];
  }

  collectPortals(children: React.ReactNode) {
    return <PortalManager portals={this.portals}>{children}</PortalManager>;
  }

  appendUniversalPortals = (html: string) {
    if (!this.portals.length) {
      return html;
    }

    const $ = load(html);
    this.portals.forEach(([children, selector]) => {
      try {
        const markup = ReactDOMServer.renderToStaticMarkup(children);
        $(markup).attr("data-react-universal-portal", "").appendTo((selector as any))
      } catch (error) {
        console.warn('Unable to render portal server-side:\n' + error.message || error.toString().split('\n')[0]);
      }
    });

    // it's important to flush one and only one time per render
    this.portals.length = 0;

    return $.html({ decodeEntities: false });
  }
}
