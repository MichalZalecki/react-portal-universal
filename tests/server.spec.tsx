import * as React from "react";
import { createUniversalPortal, flushUniversalPortals } from "../src";
import { appendUniversalPortals } from "../src/server";

describe("server-side", () => {
  beforeEach(() => {
    flushUniversalPortals();
  });

  describe("appendUniversalPortals", () => {
    it("adds static mackup from portals to rendered HTML", () => {
      const html = `
        <html>
          <head>
            <meta charset="utf-8" />
          </head>
          <body>
            <div id="root"></div>
          </body>
        </html>
      `;
      createUniversalPortal(<title>Hello, World!</title>, "head");
      createUniversalPortal(<h1>Hello, World!</h1>, "body");
      const withStaticMarkup = appendUniversalPortals(html);

      expect(withStaticMarkup.includes("<title>Hello, World!</title>"));
      expect(withStaticMarkup.includes("<h1>Hello, World!</h1>"));
      expect(withStaticMarkup).toMatchSnapshot();
    });

    it("doesn't render non-latin characters as HTML entities", () => {
      const html = `
        <html>
          <head>
            <meta charset="utf-8" />
          </head>
          <body>
            <div id="root"></div>
          </body>
        </html>
      `;
      createUniversalPortal(<title>Привет, мир!</title>, "head");
      createUniversalPortal(<h1>Привет, мир!</h1>, "body");
      const withStaticMarkup = appendUniversalPortals(html);

      expect(withStaticMarkup.includes("<title>Привет, мир!</title>"));
      expect(withStaticMarkup.includes("<h1>Привет, мир!</h1>"));
      expect(withStaticMarkup).toMatchSnapshot();
    });
  });
});
