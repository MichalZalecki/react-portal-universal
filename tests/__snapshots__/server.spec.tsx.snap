// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server-side appendUniversalPortals adds static mackup from portals to rendered HTML 1`] = `
"<html><head>
            <meta charset=\\"utf-8\\">
          <title data-react-universal-portal=\\"\\">Hello, World!</title></head>
          <body>
            <div id=\\"root\\"></div>
          
        
      <h1 data-react-universal-portal=\\"\\">Hello, World!</h1></body></html>"
`;

exports[`server-side appendUniversalPortals doesn't render non-latin characters as HTML entities 1`] = `
"<html><head>
            <meta charset=\\"utf-8\\">
          <title data-react-universal-portal=\\"\\">Привет, мир!</title></head>
          <body>
            <div id=\\"root\\"></div>
          
        
      <h1 data-react-universal-portal=\\"\\">Привет, мир!</h1></body></html>"
`;
